README.md

```diff
@@ -433,3 +434,334 @@ with the following lines.
+Copy the following hook defninitions into `pio.wake`.
+```wake
+publish dutSimCompileOptionsHooks = pioHook, loopbackHook, Nil
+
+def loopbackHook =
+  def name = "loopback"
+  def addSources = source "{blockPIOSiFiveRoot}/rtl/loopback/loopback.sv", _
+  makeBlackBoxHook name (editDUTSimCompileOptionsSourceFiles addSources)
+
+def pioHook =
+  def name = "pio"
+  def addSources = source "{blockPIOSiFiveRoot}/rtl/pio/pio.sv", _
+  makeBlackBoxHook name (editDUTSimCompileOptionsSourceFiles addSources)
+```
+
+## Making a test
+
+Currently, only c integration tests are supported. A simple test is included
+with this repository in `block-pio-sifive/tests/demo/main.c`. It looks like this.
+```c
+int main()
+{
+  // read/write to axi block
+  volatile uint32_t * pio = (uint32_t *) 0x60000;
+
+  volatile uint32_t * odata   = pio;
+  volatile uint32_t * oenable = pio + 1;
+  volatile uint32_t * idata   = pio + 2;
+
+  int fail = 0;
+  int test_len = 100;
+  for (int i = 0; i < test_len; i++) {
+    *odata = i;
+    *oenable = test_len - i;
+
+    fail |= ((*odata ^ *oenable) != *idata);
+  }
+
+  return fail;
+}
+```
+
+To tell wake how to compile this test we need to create a `TestProgramPlan`.
+`TestProgramPlan` is defined as follows:
+```wake
+tuple TestProgramPlan =
+  global Name:        String
+  global CFlags:      List String
+  global ASFlags:     List String
+  global CFiles:      List Path
+  global IncludeDirs: List String
+  global Sources:     List Path
+  global Filter:      DUTProgramCompiler => Boolean
+```
+A `TestProgramPlan` should only contain machine agnostic flags, since the
+same program plan may be run on multiple different configs. Machine-specific
+options should be filled in by the `DUTProgramCompiler`. You can specify what
+`DUTProgramCompiler`s are valid for this plan using the `Filter` field.
+
+To create a `TestProgramPlan` use `makeTestProgramPlan`. It has the following
+type signature.
+```wake
+def mySimpleProgram =
+  def programName = ${name of program} # String
+  def cfiles = ${files to compile} # List SPath
+  makeTestProgramPlan programName cfiles # Program
+```
+
+Currently, the only supported `DUTProgramCompiler` is
+`freedomMetalDUTProgramCompiler` which uses
+[freedom-metal](https://github.com/sifive/freedom-metal).
+`freedomMetalDUTProgramCompiler` provides standard libraries and will compile
+and link programs according the DTS.
+
+Since the demo test is pretty simple, we only need to specify the cfiles and the
+program name, and we can use the default parameters. Copy the following into
+`block-pio-sifive/wake/demo.wake` to create a `TestProgramPlan` for `block-pio-sifive/tests/demo/main.c`.
+```wake
+global def demo =
+  def programName = "demo"
+  def cFiles = source "{blockPIOSiFiveRoot}/tests/demo/main.c", Nil
+  makeTestProgramPlan programName cFiles
+```
+
+Suppose we did not want to hardcode the address of the pio block and we edited
+our test to look like this.
```

Comment:
Hmm, we may need to change this example, since we want to automatically generate the header file that sets the PIO address anyway.